wa/instrumentation/fps.py

```diff
@@ -0,0 +1,165 @@
+            collector_target_attr = 'package'
+        else:
+            collector_target_attr = 'view'
+        collector_target = getattr(context.workload, collector_target_attr, None)
+
+        if not collector_target:
+            self._is_enabled = False
+            msg = 'Workload {} does not define a {}; disabling frame collection and FPS evaluation.'
+            self.logger.info(msg.format(context.workload.name, collector_target_attr))
+            return
+
+        self._is_enabled = True
+        if use_gfxinfo:
+            self.collector = GfxInfoFramesInstrument(self.target, collector_target, self.period)
+            self.processor = DerivedGfxInfoStats(self.drop_threshold, filename='fps.csv')
+        else:
+            self.collector = SurfaceFlingerFramesInstrument(self.target, collector_target, self.period)
+            self.processor = DerivedSurfaceFlingerStats(self.drop_threshold, filename='fps.csv')
+        self.collector.reset()
+
+    def start(self, context):
+        if not self._is_enabled:
+            return
+        self.collector.start()
+
+    def stop(self, context):
+        if not self._is_enabled:
+            return
+        self.collector.stop()
+
+    def update_result(self, context):
+        if not self._is_enabled:
+            return
+        outpath = os.path.join(context.output_directory, 'frames.csv')
+        frames_csv = self.collector.get_data(outpath)
+        raw_output = self.collector.get_raw()
+
+        processed = self.processor.process(frames_csv)
+        processed.extend(self.processor.process_raw(*raw_output))
+        fps, frame_count, fps_csv = processed[:3]
+        rest = processed[3:]
+
+        context.add_metric(fps.name, fps.value, fps.units)
+        context.add_metric(frame_count.name, frame_count.value, frame_count.units)
+        context.add_artifact('frames', frames_csv.path, kind='raw')
+        context.add_artifact('fps', fps_csv.path, kind='data')
+        for metric in rest:
+            context.add_metric(metric.name, metric.value, metric.units)
+
+        if not self.keep_raw:
+            for entry in raw_output:
+                if os.path.isdir(entry):
+                    shutil.rmtree(entry)
+                elif os.path.isfile(entry):
+                    os.remove(entry)
+
+        self.check_for_crash(context, fps.value, frame_count.value,
+                             context.current_job.run_time.total_seconds())
+
+    def check_for_crash(self, context, fps, frames, exec_time):
+        if not self.crash_threshold:
+            return
+        self.logger.debug('Checking for crashed content.')
+        if all([exec_time, fps, frames]):
+            expected_frames = fps * exec_time
+            ratio = frames / expected_frames
+            self.logger.debug('actual/expected frames: {:.2}'.format(ratio))
+            if ratio < self.crash_threshold:
+                self.logger.error('Content for {} appears to have crashed.'.format(context.spec.label))
```

Comment:
Seems this needs to be `context.current_job.spec.label`

Reply:
Fixed